Mathlib/Algebra/Order/Monovary.lean

```diff
@@ -0,0 +1,428 @@
+      add_comm (f i • g j)]
+
+/-- Two functions antivary iff the rearrangement inequality holds. -/
+lemma antivaryOn_iff_smul_rearrangement :
+    AntivaryOn f g s ↔
+      ∀ ⦃i⦄, i ∈ s → ∀ ⦃j⦄, j ∈ s → f i • g i + f j • g j ≤ f i • g j + f j • g i :=
+  monovaryOn_toDual_right.symm.trans $ by rw [monovaryOn_iff_smul_rearrangement]; rfl
+
+/-- Two functions monovary iff the rearrangement inequality holds. -/
+lemma monovary_iff_smul_rearrangement :
+    Monovary f g ↔ ∀ i j, f i • g j + f j • g i ≤ f i • g i + f j • g j :=
+  monovaryOn_univ.symm.trans $ monovaryOn_iff_smul_rearrangement.trans $ by
+    simp only [Set.mem_univ, forall_true_left]
+
+/-- Two functions antivary iff the rearrangement inequality holds. -/
+lemma antivary_iff_smul_rearrangement :
+    Antivary f g ↔ ∀ i j, f i • g i + f j • g j ≤ f i • g j + f j • g i :=
+  monovary_toDual_right.symm.trans $ by rw [monovary_iff_smul_rearrangement]; rfl
+
+alias ⟨MonovaryOn.sub_smul_sub_nonneg, _⟩ := monovaryOn_iff_forall_smul_nonneg
+alias ⟨AntivaryOn.sub_smul_sub_nonpos, _⟩ := antivaryOn_iff_forall_smul_nonpos
+alias ⟨Monovary.sub_smul_sub_nonneg, _⟩ := monovary_iff_forall_smul_nonneg
+alias ⟨Antivary.sub_smul_sub_nonpos, _⟩ := antivary_iff_forall_smul_nonpos
+alias ⟨Monovary.smul_add_smul_le_smul_add_smul, _⟩ := monovary_iff_smul_rearrangement
+alias ⟨Antivary.smul_add_smul_le_smul_add_smul, _⟩ := antivary_iff_smul_rearrangement
+alias ⟨MonovaryOn.smul_add_smul_le_smul_add_smul, _⟩ := monovaryOn_iff_smul_rearrangement
+alias ⟨AntivaryOn.smul_add_smul_le_smul_add_smul, _⟩ := antivaryOn_iff_smul_rearrangement
+
+end LinearOrderedAddCommGroup
+
+section LinearOrderedRing
+variable [LinearOrderedRing α] {f g : ι → α} {s : Set ι} {a b c d : α}
+
+lemma monovaryOn_iff_forall_mul_nonneg :
+    MonovaryOn f g s ↔ ∀ ⦃i⦄, i ∈ s → ∀ ⦃j⦄, j ∈ s → 0 ≤ (f j - f i) * (g j - g i) := by
+  simp only [smul_eq_mul, monovaryOn_iff_forall_smul_nonneg]
+
+lemma antivaryOn_iff_forall_mul_nonpos :
+    AntivaryOn f g s ↔ ∀ ⦃i⦄, i ∈ s → ∀ ⦃j⦄, j ∈ s → (f j - f i) * (g j - g i) ≤ 0 := by
+  simp only [smul_eq_mul, antivaryOn_iff_forall_smul_nonpos]
+
+lemma monovary_iff_forall_mul_nonneg : Monovary f g ↔ ∀ i j, 0 ≤ (f j - f i) * (g j - g i) := by
+  simp only [smul_eq_mul, monovary_iff_forall_smul_nonneg]
+
+lemma antivary_iff_forall_mul_nonpos : Antivary f g ↔ ∀ i j, (f j - f i) * (g j - g i) ≤ 0 := by
+  simp only [smul_eq_mul, antivary_iff_forall_smul_nonpos]
+
+/-- Two functions monovary iff the rearrangement inequality holds. -/
+lemma monovaryOn_iff_mul_rearrangement :
+    MonovaryOn f g s ↔
+      ∀ ⦃i⦄, i ∈ s → ∀ ⦃j⦄, j ∈ s → f i * g j + f j * g i ≤ f i * g i + f j * g j := by
+  simp only [smul_eq_mul, monovaryOn_iff_smul_rearrangement]
+
+/-- Two functions antivary iff the rearrangement inequality holds. -/
+lemma antivaryOn_iff_mul_rearrangement :
+    AntivaryOn f g s ↔
+      ∀ ⦃i⦄, i ∈ s → ∀ ⦃j⦄, j ∈ s → f i • g i + f j • g j ≤ f i • g j + f j • g i := by
```

Comment:
These might as well be `mul` rather than `smul`?

Reply:
Whoops, typo indeed